flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -137,11 +156,18 @@ public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
         // if no variables are excluded (which is the default!) we re-use the general variables map
         // to save space
         return internalGetAllVariables(
-                excludedVariables.isEmpty() ? 0 : reporterIndex, excludedVariables);
+                filter,
+                excludedVariables.isEmpty() ? 0 : reporterIndex,
+                excludedVariables,
+                delimiter);
     }
 
     private Map<String, String> internalGetAllVariables(
-            int cachingIndex, Set<String> excludedVariables) {
+            CharacterFilter filter,
+            int cachingIndex,
+            Set<String> excludedVariables,
+            char delimiter) {
+        CharacterFilter newFilter = wrapWithDefaultFilter(filter, delimiter);
```

Comment:
This implies that variables returned by `parent.getAllVariables()` already had a filter applied to them for a delimiter that may not be matching `delimiter` (i.e., `registry.getDelimiter()`).
Furthermore, this also means that exclusion may not be applied correctly.

I think we need to restrict the filtering to cases where the calls originate from a reporter.

Reply:
ah yeah, thanks for pointing out ! Anyway same goes for the _AbstractMetricGroup#getMetricIdentifier()_ case no ?

Reply:
It shouldn't be a problem for `getMetricIdentifier()`; this method only relies on `this.scopeComponents` which is constant and does not fetch anything from parent groups.

Maybe I haven't explained this well.
Let's say the delimiter is `-`, the global default delimiter (returned by `registry.getDelimiter()`) is `.`, and we have a variable defined in a parent group called `-aaa.`.

The result we want is `_aaa.`.

The problem is you retrieved this variable in the first place via `parent.getAllVariables()`. This one doesn't know about your filter, and uses the delimiter returned by `registry.getDelimiter()`. It will apply this filter, resulting in `-aaa_`, and when your filter is applied we end up with `_aaa_`, which is not what we want.

The easiest way I see to solve is to only wrap the filter if the cachingIndexis greater than 0.

Reply:
thanks for the details.
It was clear already for the _getAllVariables()_ inheritance case.

Reply:
besides I don't think that testing `cachingIndex > 0` is enough to detect that the call was done by a reporter because if excluded variables are empty `cachingIndex` is set to 0 (to save memory space) even when the call is done by a reporter.

Reply:
Good catch; we have to extend that optimization to only trigger if there is neither a filter nor any excluded variables.

Reply:
what I did for clarity is that `internalGetAllVariables` takes a boolean `calledFromReporter` that is true when the provided `reporterIndex` is valid. `AbstractMetricGroup#getAllVariables()` and `AbstractMetricGroup#getAllVariables(CharacterFilter filter)` that call this method with `registry.getDelimiter()` pass `calledFromReporter == false`. Does it look good to you ?